yamux/src/connection/rtt.rs

```diff
@@ -0,0 +1,170 @@
+// Copyright (c) 2023 Protocol Labs.
+//
+// Licensed under the Apache License, Version 2.0 or MIT license, at your option.
+//
+// A copy of the Apache License, Version 2.0 is included in the software as
+// LICENSE-APACHE and a copy of the MIT license is included in the software
+// as LICENSE-MIT. You may also obtain a copy of the Apache License, Version 2.0
+// at https://www.apache.org/licenses/LICENSE-2.0 and a copy of the MIT license
+// at https://opensource.org/licenses/MIT.
+
+//! Connection round-trip time measurement
+
+use std::{
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+use parking_lot::Mutex;
+
+use crate::connection::Action;
+use crate::frame::{header::Ping, Frame};
+
+const PING_INTERVAL: Duration = Duration::from_secs(10);
+
+#[derive(Clone, Debug)]
+pub(crate) struct Rtt(Arc<Mutex<RttInner>>);
+
+impl Rtt {
+    pub(crate) fn new() -> Self {
+        Self(Arc::new(Mutex::new(RttInner {
+            rtt: None,
+            state: RttState::Initial,
+        })))
+    }
+
+    pub(crate) fn next_ping(&mut self) -> Option<Frame<Ping>> {
+        let state = &mut self.0.lock().state;
+
+        match state {
+            RttState::Initial => {}
+            RttState::AwaitingPong { .. } => return None,
+            RttState::Waiting { next } => {
+                if *next > Instant::now() {
+                    return None;
+                }
+            }
```

Comment:
This doesn't register a waker, hence we won't be measuring RTTs on idle connections, correct? Could we perhaps document this on the struct?

Reply:
:+1: c9c71b6